### ISSUE TITLE
 TotalsegmentatornRunning Slow on New M3Max Macbook Pro

### ISSUE THREAD

**Issue by QianMuXiao** [wasserth/TotalSegmentator#250]
I'm running TotalSegmentator on my M3Max chip (16-core CPU 40-core GPU) Macbook Pro and it's taking over 600 seconds to fully segment a 3D CT image of size 512*512*54, whereas it tends to take about 60 seconds on my 3070 GPU desktop. Is it possible that the latest version of totalsegmentator just uses the Macbook's CPU instead of calling the Macbook's "MPS" when using pytorch?

**Comment by francescopisu** [comment]
Pytorch's MPS backend doesn't support 3d ops yet, hence even if TotalSegmentator is doing everything right your computation would still be slow on a mac.

**Comment by QianMuXiao** [comment]
> francescopisu
Yes I used CT image data with dimensions of about 512*512*54, but before that I didn't realize that Pytorch's MPS device doesn't support 3D operations, I'll go back and re-check the docs on that, thanks a lot!

**Comment by QianMuXiao** [comment]
> Pytorch's MPS backend doesn't support 3d ops

I just checked the latest pytorch documentation it seems that the new version should now support MPS accelerated 3D operations:
https://github.com/pytorch/pytorch/pull/116580#issue-2061155133
Maybe I need to check further to be sure.

**Comment by QianMuXiao** [comment]
> Are you using TotalSegmentator for 3d segmentation ? Pytorch's MPS backend doesn't support 3d ops yet, hence even if TotalSegmentator is doing everything right your computation would still be slow on a mac.

I checked the Pytorch documentation in detail and after some practice the current night version of pytorch only supports MPS accelerated Conv3d but not ConvTranspose3d operations.

**Comment by wasserth** [comment]
It seems that quite recently pytorch finally added Conv3d to the nightly version. I pushed a commit to master to allow "mps" as device argument. I did not have a chance to test it since I am not running the newest MacOs which is required for this to work.

**Comment by wasserth** [comment]
It seems that nnUNnet also uses ConvTranspose3D which is not yet supported. So mps is not working for now.

**Comment by QianMuXiao** [comment]
> It seems that nnUNnet also uses ConvTranspose3D which is not yet supported. So mps is not working for now.

Yes, I called the mps device parameter by modifying the source code in the Totalsegmentator package, and in the latest nightly version of the pytorch environment it prompts that pytorch does not support ConvTranspose3D under mps.

**Comment by wxc-2020** [comment]
I run Totalsegmentator on  M3 MAX mbp to segment only one CT image, regardless of using --fast or --rb, the error is as follows: Background workers died. Look for the error message further up!  If there is none then your RAM was full and the worker was killed by the OS. Use fewer workers or get more RAM in that  case!

**Comment by francescopisu** [comment]
@wxc-2020 How many workers for preprocessing ? nr_thr_resamp and nr_thr_saving
My run crashed very recently because I spawned too many workers (31 to be precise) and I drained all 64 GB of RAM.

**Comment by QianMuXiao** [comment]
My running still fine，My Mac got only 48GB RAM and when I run the 3D Segment with full-res on CPU it`s slow but won`t broken and don`t need to use swap

**Comment by francescopisu** [comment]
@QianMuXiao It also depends on the specifics of your tomographic data

**Comment by QianMuXiao** [comment]
@francescopisu The MSD datasets I've been using lately seem to run segmentation fine

**Comment by wxc-2020** [comment]
@francescopisu I didn't specifically set these two parameters, which seem to be the default; In addition, my CT images are  256*256*300；I still can't find the specific reason for the error.

**Comment by francescopisu** [comment]
@wxc-2020 Can you show the entire stack trace when you get the error ?

**Comment by wxc-2020** [comment]
@francescopisu Thanks for your care. I have now uninstalled and installed all my conda and re-run Totalsegmentator, no longer report corresponding errors and run smoothly with the CPU（it takes 3 to 4 minutes for mbp with 128g memory to segment a CT image in --fast mode）.I think the reason for the error may be that there are some conflicts in my old conda env.

**Comment by QianMuXiao** [comment]
@francescopisu Your guide runs perfect on my MBP, Thanks a lot, it takes about 86s for task total  (512*512*54 CT scan from MSD spleen dataset) on my M3Max with 48G RAM.

**Comment by w1ebr** [comment]
@francescopisu I tried using your guide on my M3 MBP and when trying "install -r requirements.txt --no-cache-dir" I get an error message saying that there is no "-r" option for install (which install =  /usr/bin/install). Is there another "install" program that should be called?
Thanks!
Gene

**Comment by QianMuXiao** [comment]
@add pip before install

**Comment by QianMuXiao** [comment]
@w1ebr  add pip before install

**Comment by francescopisu** [comment]
@w1ebr My bad, I forgot the "pip" for "pip install". I updated the blog post as well. Thanks

**Comment by QianMuXiao** [comment]
@francescopisu is it possible to make my 3D slicer Using this version of pytorch？

**Comment by francescopisu** [comment]
@QianMuXiao I'm afraid that's not a trivial thing to do. I may need some support @lassoan.

**Comment by w1ebr** [comment]
Thank you!

**Comment by lassoan** [comment]
> is it possible to make my 3D slicer Using this version of pytorch

It may be just a matter of weeks until a new pytorch official version comes out that works out of the box, so probably it is not worth spending a whole lot of time with this. But, if that seems like a long time then, you can run `pip_install('https://example.com/path/to/custompytorch.tar.gz')` in Slicer's Python console to install a custom pytorch build from a URL. If the TotalSegmentator extension finds that pytorch is installed then it will use that.

**Comment by QianMuXiao** [comment]
@francescopisu maybe your quick guide commadn ‘conda create --prefix==./venv python=3.10’ should change to ‘conda create --prefix = ./venv python=3.10’？

**Comment by wxc-2020** [comment]
@francescopisu  Perfect! Thanks again!, it takes effective speed increase！

**Comment by francescopisu** [comment]
> @francescopisu maybe your quick guide commadn ‘conda create --prefix==./venv python=3.10’ should change to ‘conda create --prefix = ./venv python=3.10’？

Yes, updated.

**Comment by cutright** [comment]
Just noticed there was a discussion about this here, I also got Mac GPU working a few weeks ago: https://github.com/wasserth/TotalSegmentator/issues/39#issuecomment-2007890904

Working great for me! Thanks @wasserth et al

> FWIW, I installed the commit from the PR mentioned in the prior comment with:
pip install git+https://github.com/pytorch/pytorch.git@3c61c525694eca0f895bb01fc67c16793226051a
>
> Then set the device to 'mps' in my totalsegmentator call and it seems to work.
>
> It took about 8min to get 69 ROIs on a CT sim for a prostate case, using an Apple M2 Pro on Sonoma 14.4 and python 3.10.9. This included DICOM output. I was able to import the DICOM output into my viewer and it looks pretty good to me.

**Comment by tkerby** [comment]
Has this been integrated into the main code versions yet?

**Comment by lassoan** [comment]
TotalSegmentator developers cannot do much about this.

It is a matter of how slowly Apple is adding hardware acceleration for PyTorch - the open-source library that underpins almost all medical AI imaging tools nowadays. Apple tried to ignore PyTorch for years, expecting developers to switch proprietary Apple APIs instead. When they realized that this strategy does not work and Linux and Windows platforms are way ahead, they finally started funding the necessary developments. However, they are still investing only the absolute minimum into this, only working things that are their immediate interest, not engaging, supporting the open-source/research community, and so of course the progress is really slow.

You can just wait some more for Apple to catch up (maybe voice your dissatisfaction at the right forums to accelerate the process) or switch to platforms that better support open AI technology.

**Comment by QianMuXiao** [comment]
@lassoan actually, we only need to do the following two updates to make TotalSegmentator accelerate smoothly on the Mac version of 3D slicer:
First get a Pytorch installer that supports MPS-accelerated 3D Transconv operations and install it in the Python environment of 3Dslicer.
Then update TotalSegmentator's 3Dslicer plugin so that you can switch the device to MPS in the Mac version of the 3Dslicer plugin.
The only problem may be that this pytorch package may only accelerate TotalSegmentator but may cause bugs in other plugins.

**Comment by lassoan** [comment]
> First get a Pytorch installer that supports MPS-accelerated 3D Transconv operations 

Please work with Apple and PyTorch developers to get these in the official PyTorch binaries. If they are available there then let us know and we'll make sure they are used.

In the meantime, a few workaround options:
- A. Install a special PyTorch version manually from any source, TotalSegmentator will use it. Can you provide a complete pip install command that does this?
- B. Add an experimental option to the SlicerPyTorch extension to install a special PyTorch version on mscOS that has hardware acceleration.

**Comment by tkerby** [comment]
Apple seem to show it's now built into the nightly releases - https://developer.apple.com/metal/pytorch/
